### DISCUSSION TITLE
Nested form lists

### DISCUSSION THREAD

**Discussion by BenJeau** [mantinedev/mantine#1558] [Q&A] [answered]
Hi,

I'm trying to get nested form lists to work, but unsure how to do it or if it's even supported. Love the form list feature btw!

For example, I've tried this, but it doesn't work. With the following form:

```ts
const form = useForm({
  initialValues: {
    fields: formList([
      {
        key: randomId(),
        options: formList([
          {
            key: randomId(),
            option: "",
          },
        ]),
      },
    ]),
  },
});
```

I can iterator over the options like this:

```ts
form.values.fields[index].options.map(...)
```

But then inside the map, I can't seem to be able to call any form list function, I would assume the API would be something like (but I can see the limitation of an API designed this way for nested form list)

```ts
form.getListInputProps(
  "fields",
  index,
  "options",
  optionIndex,
  "option"
)
```

I'm thinking of maybe having nested "forms", but I feel like that can get messy quickly.

Is there a way to this currently? Thank you in advance!

> **Edit**: For a bit more context, my use case is that I'm building a web app that's a form builder, where the user can create fields and if the field has options (such as radios, checkboxes, selects), there would be options for those

**Comment by rtivital** [comment, marked as answer]
use-form currently does not support nested lists

**Reply by BenJeau** [reply]
Ok, thank you for the answer :) - I also tried doing validation between values within the same form list and that also does not seem like something currently supported. I'll look into other libs for forms

**Reply by BenJeau** [reply]
I seem to be able to simply use the `validate(values)` to validate:

```ts
useForm({ validate: (values) => {
  const errors: any = {};
  
  values.fields.forEach((field, index) => {
    ...perform validation
  });
  
  return errors;
})
```

and to just use a simple array for my nested list, since its just an array of strings. Hopefully my form won't need to be more complex, I like the simple `use-form` API of Mantine, it was easy to grasp

**Comment by Thespaxd** [comment]
I think you should look into [react-hook-form](https://react-hook-form.com) its a really powerful lightweight tool for dealing with forms 
and it supports [nested forms](https://react-hook-form.com/api/usefieldarray) as well as [controlled components](https://react-hook-form.com/api/usecontroller/controller) so its easy to integrate it with mantine

**Reply by BenJeau** [reply]
I'll look into that lib :)
